keps/sig-windows/20190714-windows-csi-support.md

```diff
@@ -0,0 +1,401 @@
+
+## Design Details
+
+### Test Plan
+
+Unit tests will be added to verify Windows related enhancements in existing Kubernetes components mentioned above.
+
+All E2E storage tests covering CSI plugins will be ported to Windows workloads and successfully executed with above enhancements in place along with csi-proxy.exe.
+
+### Graduation Criteria
+
+#### Alpha -> Beta Graduation
+
+- CSI Node Plugins for GCE PD, Azure Disk and Azure File can support Windows using csi-proxy.exe.
+- end-2-end tests in place involving CSI storage scenarios that exercise csi-proxy.exe and related enhancements.
+
+#### Beta -> GA Graduation
+
+- All in-tree storage plugins that implements support for Windows can use csi-proxy.exe along with other enhancements listed above to successfully deploy CSI plugins on Windows nodes.
+- Successful usage of csi-proxy.exe in Windows nodes by at least two storage vendors.
+
+### Upgrade / Downgrade Strategy
+
+If a cluster is downgraded from a version of Kubernetes where the plugin watcher supports Windows nodes to one that does not, existing Windows workloads that were using CSI plugins to access storage will no longer have access to the data. This loss of functionality cannot be handled in an elegant fashion.
+
+### Version Skew Strategy
+
+Different nodes in the cluster may be configured with different versions of csi-proxy.exe as part of a rolling upgrade of csi-proxy.exe. In such a scenario, it is recommended that csi-proxy.exe is upgraded first and once that is complete, the CSI Node Plugins that can take advantage of the new version of csi-proxy.exe is rolled out.
+
+Beyond the above, there are no Kubernetes version skew considerations in the context of this KEP.
+
+## Implementation History
+
+07/16/2019: Initial KEP drafted
+
+## Drawbacks
+
+The main drawback associated with the approach leveraging csi-proxy.exe is that the life cycle of that binary will need to be managed out-of-band from Kubernetes. However, cluster administrators need to maintain and manage life cycle of other core binaries like kubeproxy.exe, kubelet.exe, dockerd.exe and containerd.exe (in the future). Therefore csi-proxy.exe will be one additional binary that will need to be treated in a similar way.
+
+The API versioning scheme described above, as of now, will maintain backward compatibility indefinitely. This requires the scope of csi-proxy.exe to be limited to a very scoped down fundamental set of operations. Maintainers therefore will need to be very cautious when accepting suggestions for new APIs and enhancements. This may slow progress at times.
```

Comment:
Adding another API and maintaining is my biggest concern with the approach. (I think we've also discussed this in the @PatrickLang's doc) Does SIG storage think the API surface will be manageable and stable?

Reply:
What if instead of a common proxy we have a per-driver proxy that is maintained and deployed by each driver vendor?  It pushes the responsibility to the driver maintainers, but API and versioning compatibility is simpler since it's all under the control of one vendor.  However, this does imply we need to give CSI drivers permissions to install binaries directly on the host.

But even for Linux CSI drivers today, some already need to do this to an extent (ie iscsd doesn't work well in a container)

Reply:
At that point, why not use that same deployment mechanism to install and run the CSI driver directly on the host?

Reply:
One alternative to having a set of operations as a scoped down API exposed by the proxy is: surfacing a simple `Exec` API in the proxy that accepts arbitrary parameters and returns stdout results. That can supersede all the operations mentioned here as containers will be free to run arbitrary commands on the host. I can list this as an alternative. The main disadvantage is this makes the proxy open to massive abuse.

With this approach + bind mounting a dedicated per-plugin path in the host, plugins can drop arbitrary exes they may need (like `ebsnvme-id.exe` above) and run it through the exec interface if desired. Thoughts?